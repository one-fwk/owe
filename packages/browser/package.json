{
  "name": "@owe/browser",
  "version": "0.0.1",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "module": "lib/index-es.js",
  "files": [
    "lib"
  ],
  "scripts": {
    "prepublish": "yarn build",
    "build": "rimraf lib && rollup -c"
  },
  "dependencies": {
    "browser-detect": "^0.2.28",
    "webextension-polyfill-ts": "^0.8.9"
  }
}
