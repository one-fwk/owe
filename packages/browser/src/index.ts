export * from './browser-polyfill';
export * from './supported-browsers.enum';

export {
  Browser,
  Alarms,
  Bookmarks,
  BrowserAction,
  BrowserSettings,
  BrowsingData,
  Clipboard,
  Commands,
  ContentScripts,
  ContextualIdentities,
  Cookies,
  Devtools,
  DevtoolsInspectedWindow,
  DevtoolsNetwork,
  DevtoolsPanels,
  Dns,
  Downloads,
  Events,
  Experiments,
  Extension,
  ExtensionTypes,
  Find,
  GeckoProfiler,
  History,
  I18n,
  Identity,
  Idle,
  Management,
  Manifest,
  ContextMenus,
  Menus,
  Notifications,
  Omnibox,
  PageAction,
  Permissions,
  Pkcs11,
  Privacy,
  PrivacyNetwork,
  PrivacyServices,
  PrivacyWebsites,
  Proxy,
  Runtime,
  Search,
  Sessions,
  SidebarAction,
  Storage,
  Tabs,
  Theme,
  TopSites,
  Types,
  UserScripts,
  WebNavigation,
  WebRequest,
  Windows
} from 'webextension-polyfill-ts';