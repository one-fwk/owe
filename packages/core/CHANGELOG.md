# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.1.1](https://github.com/one-fwk/owe/compare/v0.1.0...v0.1.1) (2019-01-14)


### Bug Fixes

* **publish:** add publish access to be public ([cea649a](https://github.com/one-fwk/owe/commit/cea649a))





# 0.1.0 (2019-01-14)


### Features

* zerollup ([a2d5993](https://github.com/one-fwk/owe/commit/a2d5993))
