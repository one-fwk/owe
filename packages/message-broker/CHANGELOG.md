# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2019-01-14)


### Features

* zerollup ([a2d5993](https://github.com/one-fwk/owe/commit/a2d5993))
