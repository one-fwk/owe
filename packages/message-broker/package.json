{
  "name": "@owejs/message-broker",
  "version": "0.0.1",
  "main": "lib/index.js",
  "typings": "lib/index.d.ts",
  "module": "lib/index.mjs",
  "files": ["lib"],
  "peerDependencies": {
    "rxjs": "^6.3.3"
  },
  "dependencies": {
    "@one/core": "^1.5.1",
    "@owe/browser": "*",
    "@owe/core": "*"
  }
}
