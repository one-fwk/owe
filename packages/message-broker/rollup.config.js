import typescript from 'rollup-plugin-typescript2';
import resolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';

export default {
  input: 'src/index.ts',
  plugins: [
    typescript({
      tsconfig: 'tsconfig.build.json',
    }),
    resolve({
      jsnext: true,
      main: true,
      module: true,
    }),
    commonjs(),
  ],
  output: [
    {
      file: 'lib/index.js',
      format: 'cjs',
    },
    {
      file: 'lib/index.mjs',
      format: 'esm',
    },
  ],
};