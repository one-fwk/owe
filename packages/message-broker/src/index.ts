export * from './message-broker.module';
export * from './message-broker.service';
export * from './tokens';
export * from './services';
export * from './decorators';