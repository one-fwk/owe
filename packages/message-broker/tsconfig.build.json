{
  "extends": "../../tsconfig.base.json",
  "exclude": [
    "lib",
    "**/*.spec.ts"
  ]
}
