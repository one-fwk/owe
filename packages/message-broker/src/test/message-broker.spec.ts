import { Test, TestingModule } from '@one/testing';
import { AppContext, ContextModule } from '@owe/core';
import { browser } from '@owe/browser';
import { Subject } from 'rxjs';
import { toArray } from 'rxjs/operators';
import { MessageBrokerException } from '../message-broker.exception';
import {
  MessageBrokerModule,
  BackgroundBrokerService,
  ContentBrokerService,
  MessageBrokerService,
  PopupBrokerService,
} from '@owe/message-broker';

export function createBrokerTestingModule(context: AppContext) {
  return Test.createTestingModule({
    imports: [
      ContextModule.forRoot({ context }),
      MessageBrokerModule,
    ],
  }).compile();
}

class TestAction {
  static type = 'Whatever';

  constructor(readonly payload = 'lol') {}
}

describe('MessageBroker', () => {
  let backgroundTestingModule: TestingModule;
  let contentTestingModule: TestingModule;
  let popupTestingModule: TestingModule;

  let background: BackgroundBrokerService;
  let content: ContentBrokerService;
  let broker: MessageBrokerService;
  let popup: PopupBrokerService;

  let backgroundObserver: Subject<any>;
  let contentObserver: Subject<any>;
  let popupObserver: Subject<any>;

  let backgroundNextSpy: jest.SpyInstance;
  let contentNextSpy: jest.SpyInstance;
  let popupNextSpy: jest.SpyInstance;

  let payload: TestAction;

  beforeEach(async () => {
    backgroundTestingModule = await createBrokerTestingModule(AppContext.BACKGROUND);
    contentTestingModule = await createBrokerTestingModule(AppContext.CONTENT);
    popupTestingModule = await createBrokerTestingModule(AppContext.POPUP);

    background = backgroundTestingModule.get(BackgroundBrokerService);
    broker =  backgroundTestingModule.get(MessageBrokerService);
    content = contentTestingModule.get(ContentBrokerService);
    popup = popupTestingModule.get(PopupBrokerService);

    backgroundObserver = background.observe(TestAction);
    contentObserver = content.observe(TestAction);
    popupObserver = popup.observe(TestAction);

    backgroundNextSpy = jest.spyOn(backgroundObserver, 'next');
    contentNextSpy = jest.spyOn(contentObserver, 'next');
    popupNextSpy = jest.spyOn(popupObserver, 'next');

    payload = new TestAction();
  });

  afterEach(() => {
    backgroundNextSpy.mockRestore();
    contentNextSpy.mockRestore();
    popupNextSpy.mockRestore();
    (<any>browser.runtime).flush();
  });

  it('should throw exception when dispatching to same context', async () => {
    const error = new MessageBrokerException('Cannot dispatch action to same context');

    expect(() => background.dispatch(payload)).toThrowError(error);
  });

  it('should dispatch action to content and popup only', async () => {
    await broker.dispatch(payload).toPromise();

    expect(backgroundNextSpy).not.toHaveBeenCalled();
    expect(contentNextSpy).toHaveBeenCalledWith(payload);
    expect(popupNextSpy).toHaveBeenCalledWith(payload);
  });

  // Should the responses be an object/array where the key is the context?
  it('should return response from content and popup', async () => {
    const respondWith = ['Hello from content', 'Hello from popup'];

    content.observe(TestAction, () => respondWith[0]).subscribe();
    popup.observe(TestAction, () => respondWith[1]).subscribe();

    const responses = await broker.dispatch(payload).pipe(toArray()).toPromise();

    expect(responses).toHaveLength(2);
    expect(responses).toMatchObject(respondWith);
    expect(responses).toMatchSnapshot();
  });

  it('should dispatch action to specific context only', async () => {
    const dispatcher = background.dispatchTo(payload, AppContext.CONTENT);
    await dispatcher.toPromise();

    expect(contentNextSpy).toHaveBeenCalledWith(payload);
    expect(backgroundNextSpy).not.toHaveBeenCalled();
    expect(popupNextSpy).not.toHaveBeenCalled();
  });
});