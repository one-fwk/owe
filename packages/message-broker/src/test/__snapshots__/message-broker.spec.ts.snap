// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBroker should return response from content and popup 1`] = `
Array [
  "Hello from content",
  "Hello from popup",
]
`;
